src/runtime/managedtype.cs

```diff
@@ -8,12 +11,69 @@ namespace Python.Runtime
     /// code. It defines the common fields that associate CLR and Python
     /// objects and common utilities to convert between those identities.
     /// </summary>
+    [Serializable]
     internal abstract class ManagedType
     {
+        internal enum TrackTypes
+        {
+            Untrack,
+            Extension,
+            Wrapper,
+        }
+
+        [NonSerialized]
         internal GCHandle gcHandle; // Native handle
+
         internal IntPtr pyHandle; // PyObject *
         internal IntPtr tpHandle; // PyType *
 
+        private static readonly Dictionary<ManagedType, TrackTypes> _managedObjs = new Dictionary<ManagedType, TrackTypes>();
+
+        internal void IncrRefCount()
+        {
+            Runtime.XIncref(pyHandle);
+        }
+
+        internal void DecrRefCount()
+        {
+            Runtime.XDecref(pyHandle);
+        }
+
+        internal long RefCount
+        {
+            get
+            {
+                var gs = Runtime.PyGILState_Ensure();
+                try
+                {
+                    return Runtime.Refcount(pyHandle);
+                }
+                finally
+                {
+                    Runtime.PyGILState_Release(gs);
+                }
+            }
+        }
+
+        internal GCHandle AllocGCHandle(TrackTypes track = TrackTypes.Untrack)
+        {
+            gcHandle = GCHandle.Alloc(this);
+            if (track != TrackTypes.Untrack)
+            {
+                _managedObjs.Add(this, track);
```

Comment:
`AllocGCHandle` will only call when clr got Python's GIL, you can assume it's thread-safe, otherwise it's a bug for pythonnet internal, the access for `_managedObjs' as well.
They're internals, we should handle them internally.

Reply:
Does GIL in Python serialize type creation? It is clear it does not prevent Python code from being executed concurrently on its own.